internal/storage/inmem/store.go

```diff
@@ -114,15 +114,15 @@ func (s *Store) WriteCAS(res *pbresource.Resource, vsn string) error {
 	if existing != nil {
 		existingRes := existing.(*pbresource.Resource)
 
-		// Ensure CAS semantics.
-		if existingRes.Version != vsn {
-			return storage.ErrCASFailure
-		}
-
 		// Uid is immutable.
 		if existingRes.Id.Uid != res.Id.Uid {
 			return storage.ErrWrongUid
 		}
+
+		// Ensure CAS semantics.
+		if existingRes.Version != vsn {
+			return storage.ErrCASFailure
+		}
```

Comment:
Moving the `Uid` check first makes the error message clearer when you're trying to do a non-CAS write against an old incarnation of the resource. It's because if the resource exists but with a different `Uid`, the read will return `ErrNotFound`, so we'll leave the `Version` blank which will return `ErrCASFailure`.